### ISSUE TITLE
scientific notation is not permitted in float literals (and A-5 problem)

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#775]
Migrated from https://github.com/gpuweb/WSL/issues/345

>From JohnK's review, posted to the minutes of the recent F2F.
>https://docs.google.com/document/d/1CmKo59tjZwmePVrFpHpIG0W5shKR_GOrnNuMStPCEko/edit#bookmark=id.9yvdtpxfp241
>
> > scientific notation is missing: 1E-4 is not a valid literal

**Comment by dj2** [comment]
The spec lists a float as: `(-?[0-9].[0-9]+ | -?[0-9]+.[0-9])(e(+|-)?[0-9]+)?` which has an exponent, albeit always lowercase `e` at the moment.

**Comment by johnkslang** [comment]
Certainly different than what I reviewed.

But, it still seems to have problems.

`A-5.0` will be not be parsed as A, -, 5.0?   I think you want `-` as an operator, not part of a number.

And, I'm not seeing the . as being optional?

**Comment by johnkslang** [comment]
Perhaps the issue title should be:

"Scientific notation without a decimal not permitted, and operators incorrectly included"

**Comment by dj2** [comment]
We made a choice for the -5 to be part of the number. We've explicitly called it out in the spec:

> Note: literals are parsed greedy. This means that for statements like a -5 this will not parse as a minus 5 but instead as a -5 which may be unexpected. A space must be inserted after the - if the first expression is desired.

**Comment by johnkslang** [comment]
> literals are parsed greedy

Wait, that's not the real answer.  It's parsed as part of the number because you put it in the number.  You could also not put in the number token rule, and then _greedy_ parsing [sic] would _not_ put it in the number.

Basically, that answer is circular logic (it's part of the number because we said it was).

Further, this is not about parsing but rather about lexing/tokenizing, which happens before parsing, and I'm not just complaining about terminology...  Parsing is what you do once you create the token stream, e.g., applying numerical suffixes onto the literal's value/type, or negating the number with a `-` operator.  That is, you *tokenize* `-5.f` as `-`, `5.`, `f`, and then you *parse* those three to become the semantic meaning of `a negative floating-point 5`, providing an opportunity to give semantic errors as well.  It's kind of weirdly inconsistent that you didn't also include the suffixes as part of the number.

So, the question is not whether it's greedy, no problems there, the question is why you put the operator in the numerical token to begin with.  I'm pretty sure it should not be there.

Anyway, do we really want the following?

```
a=5f;  // okay
a = 5 f; // okay, with space before the f?  side subject
b-a;  // okay
b-5;  // error (that's pretty bizarre...)
b- 5; // okay
```

What was gained? I mean, why do this?  I only see it as a loss.

Notet that in the C++ preprocessor, spaces sometimes have semantic impact in unexpected places, and subsequently became a large source of portability issues and non-intuitive results.

**Comment by dj2** [comment]
In Tint, we _tokenize_ as `-5.`. That's why the tokenizer is greedy in that it combines the `-` with the `5.` instead of doing it the other way. The only number which currently has a suffix is unsigned int.

We put the `-` in the token because that was a choice we made in how to parse floats.

**Comment by johnkslang** [comment]
> In Tint, we tokenize as -5.. That's why the tokenizer is greedy in that it combines the - with the 5.

That has nothing to do with being greedy, it has to do with deciding that `-` is part of the number.  What's incorrect here is saying the operator is part of the number, not whether or not tokenization is greedy.  Talking about greedy totally misses the point.

The question is whether `-` is part of the number, and the answer should really be, no, that was an original oversight, which we will fix now, rather than require weird rules about using white space in the language forever more, and incorrectly having the tokenizer doing operation valuation.

**Comment by johnkslang** [comment]
So, can fixing this issue include removing the `-` from number tokenization, or do we need a separate issue?

**Comment by dj2** [comment]
It probably doesn't need a separate issue, just a PR would be fine.

**Comment by johnkslang** [comment]
Looking at the whole issue, it seems we need to change

`(-?[0-9].[0-9]+ | -?[0-9]+.[0-9])(e(+|-)?[0-9]+)?`

in a number of ways:

- no leading `-`
- support `E` as well as `e`
- make the `.` optional, if there is an `e`/`E`

It's pretty complex, which is why many parsers don't bother with a regular expression for this.  But, trying to keep the regular expression...

The first two are handled by

`([0-9].[0-9]+ | [0-9]+.[0-9])([eE](+|-)?[0-9]+)?`

The third by leveraging that to:

`([0-9].[0-9]+ | [0-9]+.[0-9])([eE](+|-)?[0-9]+)? | ([0-9]+([eE](+|-)?[0-9]+)?)`

Need to sanity check, and get group agreement.

**Comment by dneto0** [comment]
I noticed this again as part of the review to #1420, especially because we promised hex floats would have the same syntax as in C99.

I have a C99 draft spec in front of me and it has this rule:

decimal-floating-constant:
      &nbsp; &nbsp;&nbsp;fractional-constant exponent-part<sub>opt</sub> floating-suffix<sub>opt</sub> 
      &nbsp; &nbsp;&nbsp;digit-sequence exponent-part floating-suffix<sub>opt</sub> 

That second rule allows a floating point literal to not have a decimal point.  So 1e5 would be valid.

**Comment by dneto0** [comment]
I finally grokked @johnkslang's circularity argument, and I buy it.

Looking @johnkslang's last proposal, I see how it matches the grammar from the C99 spec in overall design.

Oddly, it seems to be missing `*` for zero-or-more.  

Instead of:

      ([0-9].[0-9]+ | [0-9]+.[0-9])([eE](+|-)?[0-9]+)? | ([0-9]+([eE](+|-)?[0-9]+)?)

I think this is better:

      ([0-9]*.[0-9]+ | [0-9]+.[0-9]*)([eE](+|-)?[0-9]+)? | ([0-9]+([eE](+|-)?[0-9]+)?)

                                                           ^----no decimal, but mandatory exponent

      ^--- mandatory decimal point, optional exponent--^
      which breaks into two cases
       ^--optional digits, mandatory decimal, mandatory digits
                       ^--mandatory digits, mandatory decimal, optional digits

There's redundancy between the two clauses of the mandatory-decimal case.  Following C99 we can simplify that part to:

        ([0-9]*.[0-9]+ | [0-9]+.)

So putting it together, it would be:

           ([0-9]*.[0-9]+ | [0-9]+.)([eE](+|-)?[0-9]+)? | ([0-9]+([eE](+|-)?[0-9]+)?)

As throughout, we are using `.` as itself, not as a match-any-character symbol.

**Comment by dneto0** [comment]
Fixed by #2227
